File: src/DataScience/Octave/patterns/mediator.m
function mediator()
  events = {};
  routes.inventory = @inventory_receive;
  routes.payment = @payment_receive;

  payment('paid');
  inventory('reserved');
  assert(isequal(events, {'inventory<-payment:paid', 'payment<-inventory:reserved'}));

  rejected = false;
  try
    mediator_send('payment', 'unknown', 'ignored');
  catch
    rejected = true;
  end
  assert(rejected);

  function payment(message)
    mediator_send('payment', 'inventory', message);
  end

  function inventory(message)
    mediator_send('inventory', 'payment', message);
  end

  function mediator_send(sender, recipient, message)
    if ~isfield(routes, recipient)
      error('unknown colleague: %s', recipient);
    end
    receiver = routes.(recipient);
    receiver(sender, message);
  end

  function inventory_receive(sender, message)
    events{end + 1} = ['inventory<-' sender ':' message];
  end

  function payment_receive(sender, message)
    events{end + 1} = ['payment<-' sender ':' message];
  end
end

File: src/Other/MicroPython/mediator.py
class CheckoutMediator:
    def __init__(self):
        self._routes = {}

    def register(self, name, receiver):
        self._routes[name] = receiver

    def send(self, sender, recipient, message):
        receiver = self._routes.get(recipient)
        if receiver is None:
            raise ValueError("UnknownColleague:%s" % recipient)
        return receiver(sender, message)


events = []


def inventory_receive(sender, message):
    event = "inventory<-%s:%s" % (sender, message)
    events.append(event)
    return event


def payment_receive(sender, message):
    event = "payment<-%s:%s" % (sender, message)
    events.append(event)
    return event


mediator = CheckoutMediator()
mediator.register("inventory", inventory_receive)
mediator.register("payment", payment_receive)

assert mediator.send("payment", "inventory", "paid") == "inventory<-payment:paid"
assert mediator.send("inventory", "payment", "reserved") == "payment<-inventory:reserved"
assert events == ["inventory<-payment:paid", "payment<-inventory:reserved"]

try:
    mediator.send("payment", "shipping", "dispatch")
    raise AssertionError("unknown colleague was accepted")
except ValueError as error:
    assert str(error) == "UnknownColleague:shipping"

print("MicroPython Mediator: passed")
